### ISSUE TITLE
Run Regression using darch

### ISSUE THREAD

**Issue by rz1988** [maddin79/darch#5]
Hi, 

I am trying to run a regression task using the darch package but it looks the darch class only handles classification tasks. However, I do notice there is function called minimizeAutoencoder, which trains a DBN as an autoencoder. So to run a regression I guess I can use my actual target variable instead of the input variable in an autoencoder.

My question here is 1) can I use the darch package to run regression? 2) if yes, what are the proper functions/procedures to do so? Simple examples will be greatly appreciated too.

Thanks,
Ryan

**Comment by saviola777** [comment]
Hello,

1) Yes. While darch is performing classification (i.e. it treats the network output as binary class labels) by default, by setting the parameters darch.isBin and darch.isClass to false, any darch network can be used to perform regression tasks.

2.) Assuming you are using darch 0.10.0 as released on CRAN (and not the git version), here's a simple example using the cats dataset from the MASS package:

```
library("MASS")
data(cats)

darch <- darch(Hwt ~ .,
  cats,
  normalizeWeights = T, # to avoid numerical problems with linear activation function
  scale = T, # better convergence in most cases
  layers = c(2,50,100,1), # you can play around with this, 
  darch.batchSize = 1, # higher is faster, lower converges faster (more weight updates)
  darch.learnRateWeights = .001, # needs to be very low for linear activation function
  darch.learnRateBiases = .001,
  darch.isClass = F, # this has to be false for regression tasks
  darch.isBin = F, # as does this
  darch.bootstrap = F, # set to true if you want some measure of how well the network performs on unseen examples
  darch.numEpochs = 100,
  darch.layerFunctionDefault = linearUnitDerivative) # important when not working with class labels, as the sigmoid unit function has outputs between 0 and 1
predict(darch)-cats[,3]
```

The most important parts are the ```isClass``` / ```isBin``` flags, activation functions, learn rate, and that you normalize the weights (when using linear activations). You can play around with all other parameters. Hope this helps, feel free to ask if anything is unclear. I should really extend the documentation and include stuff like this…

**Comment by rz1988** [comment]
Thank you Saviola. This saves my day.

A quick follow-up question. I like to dig deeper into the code. Do you know which file contains the darch.default method? 

Thanks,
Ryan

**Comment by saviola777** [comment]
The "interface functions" (i.e. the functions most users will use to create and train DArch instances) are currently stored in the ```interface.R``` file. I might want to change that, I'm not entirely happy with the way many files are named at the moment, suggestions for improvements are always welcome.

Also, be aware that while the release of darch 0.10.0 has been relatively recent, some parameters have been added / renamed since then, just to prepare you for possible confusion if you're looking into the current git code. Feedback and further questions are always welcome, if you're an IRC person or just comfortable with chatting in general, feel free to pay a visit to ```#bitrain``` or ```#bdk``` on ```irc.esper.net```, or via [webchat](http://chat.bdk.saviola.de). Might be a bit faster and easier than here. But I don't mind continuing the conversation here either, not too much going on here in terms of issues after all. :)

**Comment by rz1988** [comment]
Thanks Saviola777. I am new to either irc or webchat. Will explore these options later when I get back home.

Follow up questions:
1) Is it possible to set the top layer as a linear activation function and the layers below as logit activation function? Is darch.layerFunctions the right parameter to pass activation functions?

2) The darch.default function defaults to fine tune the DBN using the backpropagate function, which is based on the steepest descent approach. Is it possible to use other backpropagate functions? I think it is more appropriate to use the minimizeClassifier function because its error function is log likelihood function, whereas backpropagate uses sum squared errors.

**Comment by saviola777** [comment]
1.) Yes, assuming you have 3 layers, you'd use

```
[…]
darch.layerFunctions = list("1"=sigmoidUnitDerivative, "2"=linearUnitDerivative),
[…]
```

The indices may be a bit unintuitive, but since the first layer has no activation function (it uses the input data), the first layer function specifies the activation function for the second layer and so on. Not sure what you mean by logit activation function, but apart from the linear activation, sigmoid, hyperbolic tangent (```tanSigmoidUnitDerivative```), and softmax are available.

2.) It is possible to use other fine-tuning algorithms, at the moment rpropagation and both minimizeClassifier and minimizeAutoencoder are available, simply pass e.g.

```
[…]
darch.fineTuneFunction = minimizeClassifier,
[…]
length = 3, # example values
switchLayers = 2,
[…]
```

As seen in [this example](https://github.com/maddin79/darch/blob/b20af8f68d3e3ebab753c0da858d2a9cad16a64a/examples/example.cg.R). As for the error function, backpropagation uses

```
error <- (targetData - outputs[[numLayers]][])
```

whereas minimizeClassifier (aka Conjugate Gradients) uses

```
f = -sum(target*log(output))
```

as you stated. I have yet to fully understand this part, the backpropagation error looks wrong to me, but all attempts at changing it seem to break the algorithm (I'm not the original author of the code, just the current maintainer).

**Comment by rz1988** [comment]
Thank you Saviola777. Could you attach an example of using minimizeClassifier in a DArch instance? I can run the function on a standalone fashion but when I try to pass it to darch.fineTuneFunction it failed because minimizeClassifer needs more arguments than DArch supplies.

As to the error definition in backpropagation, I had similar question before but after some research I think it is correct. You can reference this document.

http://web.cs.swarthmore.edu/~meeden/cs81/s10/BackPropDeriv.pdf

Thanks,
Ryan

**Comment by saviola777** [comment]
As seen in my example above and in [this example](https://github.com/maddin79/darch/blob/b20af8f68d3e3ebab753c0da858d2a9cad16a64a/examples/example.cg.R), it needs two additional parameters ```length``` and ```switchLayers```. These can be passed as additional arguments to ```darch()``` and will be passed on to minimizeClassifier.

Thanks for the reference, will look into it.

**Comment by rz1988** [comment]
Thank you saviola777. The example was helpful. 

Two followup questions: 
1) After pretreain a DBN, is it possible to save its RBMs and assign the RBMs as initial weights to another DBN of the same structure?

For example, after pretrain a DBN with layers 10-100-100-2, I would like to save the pretrained weights of the DBN, W and explore different finetune algorithms on the W. Is there an easy way to do it?

2) What is the method to use to finetune a DBN? For example, if I have newed a DArch object, darch with initial weights W. What function can use to finetune this object?

Thanks,
Ryan

**Comment by saviola777** [comment]
1) The ```darch()``` function basically does up to three things: create a DArch instance, pre-train it (if rbm.numEpochs ist greater than 0), and fine-tune it (if darch.numEpochs is greater than 0). So to save the DArch instance in its state right after pre-training, simply pass darch.numEpochs=0 and you will get the DArch instance which you can then re-use for different fine-tuning scenarios, see 2).

2) There is a darch parameter in the ```darch()``` function, which you can use to pass an existing DArch instance which will then be used for training (e.g., ```darch([…], darch=mydarch)```). This way, you can use a pre-trained DArch instance and try different fine-tuning scenarios. Be aware, though, that the DArch instance is completely reset if you pass an rbm.numEpochs parameter greater than 0 -- pre-training always starts from a "clean" network state.

I hope this answers your questions.

**Comment by saviola777** [comment]
Closing after more than a month of inactivity, please create a new issue if something related comes up.
